### ISSUE TITLE
analyzing takes for too long when writing any code in flutter ! 

### ISSUE THREAD

**Issue by kivocsa99** [Dart-Code/Dart-Code#3436]
**Describe the bug**
i am on ubuntu 20.04 and iam using dart and flutter with vs code and whenever i ever write any complex  code it starts analyzing and it does that like non-stop over and over again so i have to write the code in a pad or whatever and paste it in vscode which is inconvenient at all ! 
**To Reproduce**
just by writing any complex code like `Stream<List<NewsModel>> newsStream` or using `dartz` like this `Future<Either<T1,T2>>`

**Expected behavior**
it is expected to compile and analyze it as fast as possible in every code i write . 
**Screenshots**
![image](https://user-images.githubusercontent.com/48873461/123661655-655b0280-d83d-11eb-90e6-6927b181808b.png)

**Versions :**
 - VS Code version: 1.57.1
 - Dart extension version:  3.23.1
 - Dart/Flutter SDK version: Dart : 2.13.1 /Flutter : 2.2.1

**Comment by DanTup** [comment]
Can you reliably reproduce this? If so, can you try:

- delete the problematic code from the file and Save
- enable the [analysis server instrumentation log](https://dartcode.org/docs/logging/#analyzer-instrumentation)
- restart VS Code
- wait for initial analysis to completely finish
- type the code that triggers the issue
- take a copy of that log file

If there's nothing sensitive in the log file, you can share it in a [gist](https://gist.github.com/) or email it to me at logs@dartcode.org. Otherwise, let me know and I can give you some instructions to find the useful parts from the log that might help understand what's wrong.

There were some fixes in Flutter 2.2.2 that solve issues like this, however my understanding is that they only apply to macOS, so I don't think it's that - although it may be worth upgrading to that version just to be sure.

Thanks!

**Comment by kivocsa99** [comment]
[analyze.txt](https://github.com/Dart-Code/Dart-Code/files/6727965/analyze.txt)

**Comment by kivocsa99** [comment]
@DanTup  i have also updated my flutter sdk to 2.2.2 , same bug !

**Comment by DanTup** [comment]
@kivocsa99 you did it correctly :) although in the log, all of the periods of analysis are very short (less than one second) except the initial one (6 seconds) so I think perhaps the log isn't telling the whole story.

Could you try enabling the `dart.previewLsp` setting, restart VS Code and then capture a log again? In the near future, LSP will become the default for everyone, and the log will more closely match what's seen in the UI as there is less code running in VS Code itself.

Thanks!

**Comment by kivocsa99** [comment]
@DanTup  thanks for replying , but when enabling the `dart.previewLsp` the initial analyzing for the project won't stop at all

**Comment by DanTup** [comment]
@kivocsa99 does this happen on a clean project created with `flutter create` (or **Flutter: New Application Project**) or only with this specific project?

LSP will soon be the default, so if this isn't working properly, it's very important we track it down (as the non-LSP code will be removed eventually).

**Comment by kivocsa99** [comment]
[analyze.txt](https://github.com/Dart-Code/Dart-Code/files/6732923/analyze.txt)
here is it with `dart.previewLsp` enabled , it analayzed at first but same thing happens while writing the code

**Comment by kivocsa99** [comment]
@DanTup 
i don't know if it's related or not but even in new projects when i add a new package i have to restart vs code for it to in order to
 recognize it

**Comment by DanTup** [comment]
Thanks! Unfortunately because the log is quite large, it's not very clear exactly where the issue occurred. Can you reproduce this by typing the same code into a new project? If so, could you try capturing a shorter log by doing this:

- Get the project ready to reproduce the issue (without typing the code yet)
- Run the **Dart: Capture Analysis Logs** command from the VS Code command palette
- Slowly type the code (one character at a time, and stop typing as soon as the issue is triggered)
- Click **Cancel** on the logging notification to stop logging and open the log file

Provide the log, as well as the exact code that was in the file when this occurred.

> i don't know if it's related or not but even in new projects when i add a new package i have to restart vs code in order to
recognize it

That sounds like a different issue - please open a new issue with an example/instructions and I'll take a look. I can't reproduce that myself, although a similar thing was mentioned as part of #3438 this morning.

**Comment by kivocsa99** [comment]
[analyze.txt](https://github.com/Dart-Code/Dart-Code/files/6733209/analyze.txt)
@DanTup  here is it as u requested

**Comment by DanTup** [comment]
Thanks! I can see something wrong in those logs:

```
[3:03:02 PM] [Analyzer] [Info] <== {"method":"$/progress","params":{"token":"ANALYZING","value":{"kind":"begin","title":"Analyzing…"}},"jsonrpc":"2.0"}
[3:03:03 PM] [Analyzer] [Info] <== {"method":"$/progress","params":{"token":"ANALYZING","value":{"kind":"begin","title":"Analyzing…"}},"jsonrpc":"2.0"}
```

It appears that the analysis event is sent twice which is unexpected (it should always `end` the first one before starting another). I'll do some more digging to see if I can trigger it locally.

**Comment by kivocsa99** [comment]
@DanTup this is what iam talking about it appears that it is a endless "analyzing" loop , so what's wrong here and why is that happening ?

**Comment by DanTup** [comment]
I've filed https://github.com/dart-lang/sdk/issues/46505 about this as this appears to be a parser issue. I'll close this one in favour of there, as it's not something that can be fixed in the VS Code extension. Thanks!
